File: packages/engine/src/lifecycle/cleanup.ts
import { getContext } from "iii-sdk";
import type { StateKV } from "../state/kv.js";
import { SCOPES } from "../state/schema.js";
import { getDocker } from "../docker/client.js";
import type { Sandbox } from "../types.js";

export async function cleanupAll(kv: StateKV): Promise<void> {
  const sandboxes = await kv.list<Sandbox>(SCOPES.SANDBOXES);
  const docker = getDocker();

  for (const sandbox of sandboxes) {
    try {
      const container = docker.getContainer(`iii-sbx-${sandbox.id}`);
      await container.stop().catch(() => {});
      await container.remove({ force: true });
    } catch (err: any) {
      getContext().logger.warn("Cleanup failed for container", {
        id: sandbox.id,
        error: err?.message,
      });
    }
    await kv.delete(SCOPES.SANDBOXES, sandbox.id);
  }
}

File: packages/worker/src/lifecycle/cleanup.rs
use std::sync::Arc;
use tracing::{info, warn};

use crate::runtime::SandboxRuntime;
use crate::state::{scopes, StateKV};
use crate::types::Sandbox;

pub async fn cleanup_all(rt: &Arc<dyn SandboxRuntime>, kv: &StateKV) {
    let sandboxes: Vec<Sandbox> = kv.list(scopes::SANDBOXES).await;

    for sandbox in &sandboxes {
        let cn = format!("iii-sbx-{}", sandbox.id);
        let stop_result = rt.stop_sandbox(&cn).await;
        if stop_result.is_err() {
            warn!(id = %sandbox.id, "Stop failed during cleanup");
        }
        let remove_result = rt.remove_sandbox(&cn, true).await;
        if let Err(e) = remove_result {
            warn!(id = %sandbox.id, error = %e, "Remove failed during cleanup");
        }
        let _ = kv.delete(scopes::SANDBOXES, &sandbox.id).await;
    }

    if !sandboxes.is_empty() {
        info!(count = sandboxes.len(), "Cleanup complete");
    }
}
